### PR TITLE
`1930` local storage version control

### PR DESCRIPTION
**Pull request by Da-Colon**
Noticed that I missed the Average Block Time so added that here.

### CODE REVIEW

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 1]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
...
+    case CacheKeys.MASTER_COPY:
+      return MASTER_CACHE_VERSION;
+    case CacheKeys.PROPOSAL_ARCHIVE:
+      return PROPOSAL_CACHE_VERSION;
+    case CacheKeys.AVERAGE_BLOCK_TIME:
+      return AVERAGE_BLOCK_TIME_CACHE_VERSION;
+    default:
+      throw new Error('Invalid cache name');
+  }
+};
```
Comment (Da-Colon) [inline src/hooks/utils/cache/cacheDefaults.ts:123]: Is this kinda what you are thinking?  @adamgall

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 2]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
...
+export const FAVORITES_CACHE_VERSION = 1;
+export const PROPOSAL_CACHE_VERSION = 1;
+export const AVERAGE_BLOCK_TIME_CACHE_VERSION = 1;
+
+export const getCacheVersion = (cacheName: CacheKeys): number => {
+  switch (cacheName) {
+    case CacheKeys.FAVORITES:
+      return FAVORITES_CACHE_VERSION;
+    case CacheKeys.MASTER_COPY:
+      return MASTER_CACHE_VERSION;
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts]: Good call

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 3]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
...
+    default:
+      throw new Error('Invalid cache name');
+  }
+};
 
 /**
  * Cache default values.
  *
  * Cache keys are not required to have a default value.
+ * @todo: The CACHE_DEFAULTs seem to be used for indexDB, But favorites is localstorage. We will need to revisit this.
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts:129]: Yeah 100% Another series can take a look at cleaning up indexDB in the future

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 4]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
...
+export const PROPOSAL_CACHE_VERSION = 1;
+export const AVERAGE_BLOCK_TIME_CACHE_VERSION = 1;
+
+export const getCacheVersion = (cacheName: CacheKeys): number => {
+  switch (cacheName) {
+    case CacheKeys.FAVORITES:
+      return FAVORITES_CACHE_VERSION;
+    case CacheKeys.MASTER_COPY:
+      return MASTER_CACHE_VERSION;
+    case CacheKeys.PROPOSAL_ARCHIVE:
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts]: updated.

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 5]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
   [key: string]: any;
 }
 
-export const CURRENT_CACHE_VERSION = 1;
+export const MASTER_COPY_CACHE_VERSION = 1;
+export const FAVORITES_CACHE_VERSION = 1;
+export const PROPOSAL_CACHE_VERSION = 1;
+export const AVERAGE_BLOCK_TIME_CACHE_VERSION = 1;
```
Comment (adamgall) [inline src/hooks/utils/cache/cacheDefaults.ts:108]: Do these need to be exported?

Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts:108]: fair

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 6]
```diff
@@ -96,12 +102,31 @@ interface IndexedObject {
...
+    case CacheKeys.MASTER_COPY:
+      return MASTER_COPY_CACHE_VERSION;
+    case CacheKeys.PROPOSAL_CACHE:
+      return PROPOSAL_CACHE_VERSION;
+    case CacheKeys.AVERAGE_BLOCK_TIME:
+      return AVERAGE_BLOCK_TIME_CACHE_VERSION;
+    default:
+      throw new Error('Invalid cache name');
+  }
+};
```
Comment (DarksightKellar) [inline src/hooks/utils/cache/cacheDefaults.ts:123]: Could you combine above into sth like a constant `mapCacheNameToVersion`?

File: src/hooks/utils/cache/cacheDefaults.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (DarksightKellar) [reply src/hooks/utils/cache/cacheDefaults.ts:129]: Seconded. Uh, thirded?

**Review by DarksightKellar — CHANGES_REQUESTED** [review]

File: src/hooks/utils/cache/cacheDefaults.ts (hunk 6 above)
Comment (adamgall) [reply src/hooks/utils/cache/cacheDefaults.ts:123]: @DarksightKellar I don't follow what you're suggesting here...

**Review by adamgall — CHANGES_REQUESTED** [review]
`export enum CacheKeysV0 { ...` looks unused, can remove?

**Comment by Da-Colon** [comment]
> `export enum CacheKeysV0 { ...` looks unused, can remove?

I was going keeping this around for the next PR's migration. But I don't think I really need it since I'm only going to migrate the Favorites. I'll remove this in the next PR.

File: src/hooks/utils/cache/cacheDefaults.ts (hunk 6 above)
Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts:118]: You mean like this? (Taking a guess)
```js
const CACHE_VERSIONS: { [key: string]: number } = {
  [CacheKeys.FAVORITES]: 1,
  [CacheKeys.MASTER_COPY]: 1,
  [CacheKeys.PROPOSAL_CACHE]: 1,
  [CacheKeys.AVERAGE_BLOCK_TIME]: 1,
};

export const getCacheVersion = (cacheName: CacheKeys): number => {
  const version = CACHE_VERSIONS[cacheName];
  if (version === undefined) {
    throw new Error('Invalid cache name');
  }
  return version;
};
```

(In lastest commit)

Comment (DarksightKellar) [reply src/hooks/utils/cache/cacheDefaults.ts:118]: I was thinking more like:

```
export const CACHE_VERSIONS = {
  [CacheKeys.FAVORITES]: 1,
  [CacheKeys.MASTER_COPY]: 1,
  [CacheKeys.PROPOSAL_CACHE]: 1,
  [CacheKeys.AVERAGE_BLOCK_TIME]: 1,
};


// elsewhere:
const favsCacheVersion = CACHE_VERSIONS[CacheKeys.FAVORITES]
```

So we can get rid of `getCacheVersion` and not have to do dynamic error checking, as inferred typing will prevent `CACHE_VERSIONS['invalidKey']` in the first place.

Comment (Da-Colon) [reply src/hooks/utils/cache/cacheDefaults.ts]: I was sooo close ha. Coolio updated and pushed sir.

**Review by DarksightKellar — APPROVED** [review]
Noice!

**Review by adamgall — APPROVED** [review]

**Pull request merged**